apps/emqx/src/emqx_limiter/src/emqx_limiter_client.erl

```diff
@@ -0,0 +1,90 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2025 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_limiter_client).
+
+-include("logger.hrl").
+
+-export([
+    new/2,
+    try_consume/2,
+    put_back/2
+]).
+
+-type state() :: term().
+
+-type t() :: #{
+    module := module(),
+    state := state()
+}.
+
+-export_type([t/0, state/0]).
+
+%%--------------------------------------------------------------------
+%% Callbacks
+%%--------------------------------------------------------------------
+
+-callback try_consume(state(), non_neg_integer()) -> {boolean(), state()} | boolean().
+
+-callback put_back(state(), non_neg_integer()) -> state().
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+-spec new(module(), state()) -> t().
+new(Module, State) ->
+    #{module => Module, state => State}.
+
+-spec try_consume(t(), non_neg_integer()) -> {boolean(), t()}.
+try_consume(#{module := Module, state := State} = Limiter, Amount) ->
+    try Module:try_consume(State, Amount) of
+        {Result, NewState} when is_boolean(Result) ->
+            {Result, Limiter#{state := NewState}};
+        Result when is_boolean(Result) ->
+            {Result, Limiter}
+    catch
+        error:Reason ->
+            ?SLOG_THROTTLE(
+                error,
+                #{
+                    msg => failed_to_consume_from_limiter,
+                    reason => Reason,
```

Comment:
Q: how to recover from this error ?

Reply:
>how to recover from this error ?
Create limiters back, e.g. by recreating the listener,

Personally, I'd prefer to just propagate error here to avoid inconsistent state.

The workaround comes from https://github.com/emqx/emqx/blob/master/apps/emqx/test/emqx_listeners_SUITE.erl#L59
where there is a scenario where we may "fail" to stop a listener and stop it again.

Reply:
So, the recovery is to stop a corresponding listener and start it back.

Reply:
Maybe add the possible cause and recover to the log.
e.g. `cause => "limiter not initialized", hint => "Try restart the listener"`

also, should log the limiter identifier too ?